backend/src/routes/upload.ts

```diff
@@ -38,52 +40,75 @@ export class UploadRoutes {
+    const isModelSubmission = 'model' in body
+    const isInstrumentSubmission = 'instrument' in body
+
     let instrument, model
+    let uploadRepo: Repository<InstrumentUpload | ModelUpload>
+    let productRepo: Repository<RegularFile | ModelFile>
 
     const site = await this.siteRepo.findOne(req.params.site)
     if (site == undefined) {
-      return next({ status: 422, errors: 'Unknown site'})
+      return next({status: 422, errors: 'Unknown site'})
     }
 
-    let uploadRepo: Repository<InstrumentUpload|ModelUpload>
-    if ('instrument' in body && 'model' in body) return next({ status: 422, errors: 'Both "instrument" and "model" fields may not be specified'})
-    if ('instrument' in body) {
+    if (isModelSubmission && isInstrumentSubmission) return next({
+      status: 422,
+      errors: 'Both "instrument" and "model" fields may not be specified'
+    })
+
+    if (isInstrumentSubmission) {
       instrument = await this.instrumentRepo.findOne(body.instrument)
-      if (instrument == undefined) return next({ status: 422, errors: 'Unknown instrument'})
+      if (instrument == undefined) return next({status: 422, errors: 'Unknown instrument'})
       uploadRepo = this.instrumentUploadRepo
-    }
-    else if ('model' in body) {
+      productRepo = this.regularFileRepo
+    } else if (isModelSubmission) {
       model = await this.modelRepo.findOne(body.model)
-      if (model == undefined) return next({ status: 422, errors: 'Unknown model'})
+      if (model == undefined) return next({status: 422, errors: 'Unknown model'})
       uploadRepo = this.modelUploadRepo
-    }
-    else return next({status: 422, errors: 'Metadata must have either "instrument" or "model" field'})
+      productRepo = this.modelFileRepo
+    } else return next({
+      status: 422,
+      errors: 'Metadata must have either "instrument" or "model" field'
+    })
 
     // Remove existing metadata if its status is created
-    const existingCreatedMetadata = await uploadRepo.findOne({checksum: body.checksum, status: Status.CREATED})
+    const existingCreatedMetadata = await uploadRepo.findOne({
+      checksum: body.checksum,
+      status: Status.CREATED
+    })
     if (existingCreatedMetadata != undefined) {
       await uploadRepo.remove(existingCreatedMetadata)
     }
 
     const existingUploadedMetadata = await uploadRepo.findOne({checksum: body.checksum})
     if (existingUploadedMetadata != undefined) {
-      return next({ status: 409, errors: 'File already uploaded' })
+      return next({status: 409, errors: 'File already uploaded'})
     }
 
-    const allowUpdate = ('allowUpdate' in body && body.allowUpdate.toString().toLowerCase() === 'true')
+    let allowUpdate = ('allowUpdate' in body && body.allowUpdate.toString().toLowerCase() === 'true')
 
-    // With allowUpdate flag, keep existing uuid to avoid duplicate files
+    const isExistingFreezedProduct = await this.isFreezedProduct(body, productRepo)
```

Comment:
Kun toteutetaan provenanssi LV1/LV0 fileille, niin halutaan varmaan PUT /files routessa asettaa upload-taulun allowUpdate-flagi falseksi payloadin mukana tulleen raakafileen id:n perusteella, näin vältytään tämän routen if-härpäkkeiltä, ja upload-taulun tila vastaa paremmin todellisuutta (allowUpdate aina false, jos filettä ei voi enää päivittää). Ei vaadi toimenpiteitä nyt, pohdintaa tulevan varalle.

Reply:
Totta, tähän voidaan palata.